go/vt/topo/srv_keyspace.go

```diff
@@ -115,6 +124,403 @@ func (ts *Server) GetSrvKeyspaceNames(ctx context.Context, cell string) ([]strin
+	if err != nil {
+		return nil, err
+	}
+
+	wg := sync.WaitGroup{}
+	rec := concurrency.AllErrorRecorder{}
+	servingCells = make([]string, 0)
+	var mu sync.Mutex
+	for _, cell := range cells {
+		wg.Add(1)
+		go func(cell, keyspace string) {
+			defer wg.Done()
+			srvKeyspace, err := ts.GetSrvKeyspace(ctx, cell, si.keyspace)
+			switch {
+			case err == nil:
+				for _, partition := range srvKeyspace.GetPartitions() {
+					for _, shardReference := range partition.ShardReferences {
+						if shardReference.GetName() == si.ShardName() {
+							func() {
+								mu.Lock()
+								defer mu.Unlock()
+								// Check that this cell hasn't been added already
+								for _, servingCell := range servingCells {
+									if servingCell == cell {
+										return
+									}
+								}
+								servingCells = append(servingCells, cell)
+							}()
+						}
+					}
+				}
+			case IsErrType(err, NoNode):
+				// NOOP
+				return
+			default:
+				rec.RecordError(err)
+				return
+			}
+		}(cell, si.Keyspace())
+	}
+	wg.Wait()
+	if rec.HasErrors() {
+		return nil, NewError(PartialResult, rec.Error().Error())
+	}
+	return servingCells, nil
+}
+
+// GetShardServingTypes returns served types for given shard across all cells
+func (ts *Server) GetShardServingTypes(ctx context.Context, si *ShardInfo) (servingTypes []topodatapb.TabletType, err error) {
+	cells, err := ts.GetCellInfoNames(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	wg := sync.WaitGroup{}
+	rec := concurrency.AllErrorRecorder{}
+	servingTypes = make([]topodatapb.TabletType, 0)
+	var mu sync.Mutex
+	for _, cell := range cells {
+		wg.Add(1)
+		go func(cell, keyspace string) {
+			defer wg.Done()
+			srvKeyspace, err := ts.GetSrvKeyspace(ctx, cell, si.keyspace)
+			switch {
+			case err == nil:
+				func() {
+					mu.Lock()
+					defer mu.Unlock()
+					for _, partition := range srvKeyspace.GetPartitions() {
+						partitionAlreadyAdded := false
+						for _, servingType := range servingTypes {
+							if servingType == partition.ServedType {
+								partitionAlreadyAdded = true
+								break
+							}
+						}
+
+						if !partitionAlreadyAdded {
+							for _, shardReference := range partition.ShardReferences {
+								if shardReference.GetName() == si.ShardName() {
+									servingTypes = append(servingTypes, partition.ServedType)
+									break
+								}
+							}
+						}
+
+					}
+				}()
+			case IsErrType(err, NoNode):
+				// NOOP
+				return
+			default:
+				rec.RecordError(err)
+				return
+			}
+		}(cell, si.Keyspace())
+	}
+	wg.Wait()
+	if rec.HasErrors() {
+		return nil, NewError(PartialResult, rec.Error().Error())
+	}
+	return servingTypes, nil
+}
+
+// UpdateSrvKeyspacePartitions will make sure the disableQueryService is
+// set appropriately in tablet controls in srvKeyspace.
+func (ts *Server) UpdateSrvKeyspacePartitions(ctx context.Context, keyspace string, shards []*ShardInfo, tabletType topodatapb.TabletType, cells []string, remove bool) (err error) {
```

Comment:
Do you think this function would be simpler if it was split into two: add and remove?

Reply:
I think so. Refactored that way. Let me know your thoughts.